src/transformers/models/llama/modeling_llama.py

```diff
@@ -1102,6 +1064,7 @@ def _prepare_4d_causal_attention_mask_with_cache_position(
 
 class LlamaForCausalLM(LlamaPreTrainedModel, GenerationMixin):
     _tied_weights_keys = ["lm_head.weight"]
+    _tp_plan = {"lm_head": "colwise_rep"}
```

Comment:
Trying to think out loud here, 

Do we want keep all the tp_plans as part of the config classes? We pull them here from config as needed, WDYT?

My thinking goes like adding to the LlamaConfig something like

```
llama_for_causal_lm_addon_tp_plan = {"lm_head": "colwise_rep"}
```

So here in the actual model class we pull in like we are pulling in for LlamaModel class (decoder layers)

Reply:
That's a good question. The setting in the config class in current PR is a bit conservative so as to align with the config's definition:
```
class LlamaConfig(PretrainedConfig):
    This is the configuration class to store the configuration of a [`LlamaModel`].
```

Reply:
Actually up to debate.
The thing is, `LlamaConfig` is used to init all `LlamaPreTrainedModel`, thus warrants that you can have a need for multiple extra keys (add ons). As such, I think it makes more sense to have them in the class itself. 
But, indeed confining to the config would also be easier. I am just worried about extra add ons that would be "visible" when you have a model that is unrelated (like a SequenceClassification model). 

We can fetch the `_tp_plan` from the class on post init or `super().__init__(config)`

Reply:
> I think it makes more sense to have them in the class itself

Trying to understand, when we confine plans entirely to the model class instead of config classes, how does it allow users to modify tp plans through `config.json`?

An orthogonal thought, do you think we should instead introduce separate config file like we are doing for `FSDP config` where we pass in `fsdp_config` and here it could be `tp_config`?

Reply:
Ah the thing is, we might not want to `serialize` it in the config.json, but they can pass a TP-Plan when initializing as kwargs

Reply:
Yeah we could also think about a separate file, BUT I think going for something close to our `device_map` will be easier. If there is a strong push, then we can have a config as well!

Reply:
I see @ArthurZucker, I thought we wanted to use config.json to both serialise and deserialise for users motivated from your comment - https://github.com/huggingface/transformers/issues/32470#issuecomment-2275768887 :)